rholang/src/main/k/rholang.k

```diff
@@ -1,132 +1,1039 @@
+     <GlobalListofInIds> L </GlobalListofInIds> [outcreation]
+
+rule <thread> <k> for(D:Chans <- C){ Q:HigherProc } => . </k>
+       (.Bag => <in> for(D:Chans <- C){ Q:HigherProc } </in> )
+       (.Bag => <InData> I ; L </InData> )
+    </thread>
+    <NewIntGenerator> I => I +Int 1 </NewIntGenerator>
+    <GlobalListofOutIds> L </GlobalListofOutIds>
+    <GlobalListofInIds> ... .List => ListItem(I) ... </GlobalListofInIds> [increation]
+
+// Verify the ID of one is in the list of the other
+// If they are, combine the two threads and begin
+// checking if there is a match.
+// (a) the receive's ID is in the list of the send
+rule
+    // Delete this thread, to abosrbed it into the other thread
+    // (we do this by emptying all the cells, which rewrites to . )
+    <thread>
+        <out> C:Chan!(P:HigherProcs) => . </out>
+        <OutData> I:Int ; L:List => . </OutData>
+     </thread>
+     // Combine the two threads
+     <thread>
+        <in> for(D:Chans <- C){ Q:HigherProc } </in>
+        <InData> J ; ListHead:List (ListItem(I) => .) ListTail:List </InData>
+        // Import all the information from the out-cell
+        (.Bag => <out> C:Chan!(P:HigherProcs) </out>)
+        (.Bag => <OutData> I ; L </OutData>)
+        // Begin checking the match in the k-cell
+        (.Bag => <k> listmatch( P ; D ) </k>)
+        // Keep the body of the listen statement to rewrite as we go along
+        (.Bag => <subst> Q </subst>)
+      </thread>
+
+// (b) the send's ID is in the list of the receive
+rule <thread>
+       <out> C:Chan!(P:HigherProcs) </out>
+       <OutData> J ; ListHead:List (ListItem(I) => .) ListTail:List </OutData>
+       // Absorb all the information from the in-thread
+       (.Bag => <in> for(D:Chans <- C){ Q:HigherProc } </in>)
+       (.Bag => <InData> I:Int ; L:List </InData>)
+       // Begin checking the match in the k-cell
+       (.Bag => <k> listmatch( P ; D ) </k>)
+       // Keep the body of the listen statement to rewrite as we go along
+       (.Bag => <subst> Q </subst>)
+    </thread>
+    // Delete this thread, as we abosrbed it into the other thread
+    // (we do this by emptying all the cells, which rewrites to . )
+    <thread> ...
+       <in> for(D:Chans <- C){ Q:HigherProc } => . </in>
+       <InData> I ; L => . </InData>
+    </thread>
+
+
+// If there is ever a $(MATCHFAIL), we have not matched and we separate the thread
+// into their original in- and out-threads.
+rule <thread> ...
+        // The trigger
+        <k> #(MATCHFAIL) ~> _ => . </k>
+        // Check to make sure the receive is not persistent
+        <in> for(D:Chans <- C){ Q:HigherProc } </in>
+        // Keep the in-thread intact, deleting everything else
+        <out> O => . </out>
+        <OutData> S:Int ; L:List => . </OutData>
+        <subst> B => . </subst>
+    ... </thread>
+    // Recreate the out-thread; these will not try to match each other again
```

Comment:
Just out of curiosity: how are these prevented from matching again?

Reply:
In the <InData> and <OutData> cells is a list of cells to which it could potentially match; when the match initiates, the respective id numbers in each list are removed so they don't try to match again.